### PR TITLE
feat: optimize github/gitee token process

### PR DESCRIPTION
**Pull request by frank-zsy**
## Brief Information

This pull request is in the type of ([more info](https://github.com/angular/angular/blob/main/CONTRIBUTING.md#type) about types):

- [ ] build
- [ ] ci
- [ ] docs
- [x] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] test

Related issues ([all available keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):

- 

## Details

This PR will optimize the GitHub token retrieve process. Originally, we requires users to generate the token manually and set into the HyperCRX settings, this is not friendly enough for some users especially non-tech users.

Now, we are using HyperCRX GitHub OAuth App to do this, user only need to click the button and HyperCRX will redirect to GitHub OAuth page for users to authorize HyperCRX App to get personal token automatically. Although this requires an OAuth server to get the token and get back to users, the process is much more easy for users.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)

### CODE REVIEW

**Comment by frank-zsy** [comment]
Currently, we only has a very basic version which will log the information into console but all is good now.

@wangyantong2000 Could you keep working on this PR and make it available for production?

File: src/pages/Options/components/GitHubToken.tsx [hunk 1]
```diff
@@ -20,6 +20,54 @@ const GitHubToken = () => {
...
+        interactive: true,
+      },
+      function (redirectUrl) {
+        if (chrome.runtime.lastError || !redirectUrl) {
+          console.error(chrome.runtime.lastError ? chrome.runtime.lastError.message : 'Authorization failed.');
+          return;
+        }
+        const code = new URL(redirectUrl).searchParams.get('code');
+        console.log('[FastPR]: Get session code:', code?.slice(0, 2).padEnd(code.length - 2, '*'));
+        fetch('http://8.147.129.123/github', {
```
Comment (frank-zsy) [inline src/pages/Options/components/GitHubToken.tsx:42]: This URL will be replaced as `https://oauth.hypercrx.cn/github` in the future after the domain is valid in few days.

**Comment by frank-zsy** [comment]
I also added the Gitee token process. Gitee is a little bit different than GitHub, GitHub OAuth App will return a permanent token without expire date ( maybe, as far as I know).

But Gitee will return an object with token and `expire_in` and `refresh_token`, if the token is expired, we can directly refresh the token with `refresh_token` from HyperCRX without auth or connect to server.

The doc is here: https://gitee.com/api/v5/oauth_doc#/

**Comment by wangyantong2000** [comment]
The current issues with running this part locally are as follows.

image
image

Moreover, in the Edge browser, ` chrome.identity` seems to be ineffective, and Edge may not have this API, which is `null`.

**Comment by wangyantong2000** [comment]
Sorry, after updating the Edge browser, `chrome. identity` is also valid.

**Comment by frank-zsy** [comment]
The `redirect_uri` should be fixed for OAuth App, so I changed the procedure, the OAuth App will directly redirect to server side and server will get the token and redirect to extension so extension can get the token from the URL.

**Comment by frank-zsy** [comment]
@wangyantong2000 I think I've finished all the functionality of new token retrieval procedure. You can test it in your local env and report any bug you find, and feel free to optimize the code as you like.

And the PR will not be ready until hypercrx.cn domain is valid.

**Comment by frank-zsy** [comment]
@Xsy41 Could you also test this functionality?

**Comment by Xsy41** [comment]
Good work! I think it's a great feature.

**Comment by Xsy41** [comment]
Whether user consent is required for each binding？There's nothing wrong with the function.

**Comment by frank-zsy** [comment]
> Whether user consent is required for each binding?

No, the authorization action will be only required for the first time. Actually this action means user authorizes the HyperCRX OAuth App to access his account. So if the user finish the authorization in first time, he will not need to consent again even in a different device. So after you unbind the account, re-bind account is totally automated and there will not be a pop-up page.

But there will be pop-up page in following situations:

- User need to login. So if you did not login GitHub/Gitee in current browser, the page will pop up and requires login action in the first place. And if you logout GitHub/Gitee, bind action will requires login too.
- User did not authorize the App. So after login, authorization page will pop up for user to finish the authorization with all the setting details.
- The authorization revoked. After the initial authorization, both users and the App could revoke the authorization in GitHub/Gitee, so users need to authorize again after revoke.
- The App setting updated. If the OAuth App update its setting, like the scope of the token or callback URL, this means the App is different than the former one, so users need to authorize again with the new settings.

Except the above situations, if users keep login on GitHub/Gitee, the bind action will be fully automated after the first time.

**Comment by Xsy41** [comment]
This setup sounds user-friendly, especially with automated re-binding in most cases. Thanks for your thorough response!

**Comment by frank-zsy** [comment]
The hypercrx.cn domain has been ready and already changed the online URL, this PR is ready for merge now. @wangyantong2000

**Comment by wangyantong2000** [comment]
OK

**Comment by wangyantong2000** [comment]
@Xsy41 After the refactoring work on the options page is merged, you can update and modify the comments on the token related part of the FastPR process in [911](https://github.com/hypertrons/hypertrons-crx/issues/911) again～

**Pull request merged**